hardware/VirtualThermostat.cpp

```diff
@@ -0,0 +1,740 @@
+/* 060 % */ { 1,1,1,1,1,1,0,0,0,0 },
+/* 070 % */ { 1,1,1,1,1,1,1,0,0,0 },
+/* 080 % */ { 1,1,1,1,1,1,1,1,0,0 },
+/* 090 % */ { 1,1,1,1,1,1,1,1,1,0 },
+/* 100 % */ { 1,1,1,1,1,1,1,1,1,1 },
+                    
+};                  
+
+//compute the thermostat output switch value 
+// the output is modulated in a time period of 10 minute
+// each minute , the output is activated dependeing time and percent.
+//input : Min : the minute counter 0..59
+//input : PowerPercent : the power modulation  0..100%
+//output 1 : activated
+
+short VirtualThermostat::ComputeThermostatOutput ( int Min , int PowerPercent )
+{
+  if (PowerPercent>100) PowerPercent=100;
+  //get row value for ThermostatOutput
+  PowerPercent = (PowerPercent  ) / MODULATION_STEP ;
+  short switchValue = ThermostatOutput[PowerPercent][Min%MODULATION_DURATION]; 
+  return switchValue;
+}                    
+  
+//return the poqer modulation in function of Room , Exterior and Target Temperature,
+int VirtualThermostat::ComputeThermostatPower ( int index , double RoomTemp , double TargetTemp , double ExteriorTemp , double CoefProportional , double CoefIntegral , double CoefDerivated  )
+{
+    (void)CoefDerivated;
+    (void)ExteriorTemp;
+
+  int PowerModulation = 0;
+  double DeltaTemp = TargetTemp-RoomTemp;
+  CircularBuffer * Delta = DeltaTemps[index];
+  if (Delta==0)
+  {
+    Delta= new CircularBuffer(INTEGRAL_DURATION);
+    DeltaTemps[index] = Delta;
+  }
+  //put last value
+  Delta->Put(DeltaTemp);
+  //need heating
+  if (DeltaTemp>0)
+  {
+    //PID regulation
+    // coef integral is the sum of delta temperature since last 10 minutes.
+	  PowerModulation = int(DeltaTemp * CoefProportional + Delta->GetSum() * CoefIntegral /INTEGRAL_DURATION );
+  }
+	else
+		Delta->Clear(); //clear integral part
+
+  if (PowerModulation>100) PowerModulation=100;
+  if (PowerModulation<0  ) PowerModulation=0;
+  return PowerModulation;
+}                    
+
+//return the output command & level from cmd value
+//cmd ='On;Off;Set Level XX "
+void getCommand(std::string &Cmd, std::string &OutCmd, int &level)
+{
+	OutCmd = Cmd; level = 0;
+
+	if (Cmd == "Off"){
+		level = 0;
+	}
+	else if (Cmd == "On") {
+		level = 100;
+	}
+	else {
+		std::vector<std::string> results;
+		StringSplit(Cmd, " ", results);
+        if (results.size()==3){
+		    if ( (results[0] == "Set") && (results[1] == "Level")) {
+			    OutCmd = "Set Level"; 
+                try{
+                    level = std::stoi (results[2] ) ;
+                }
+                catch(...){
+                    _log.Log(LOG_ERROR,"Invalid Switch command : %s ", Cmd.c_str() );
+                }
+            }
+            else
+ 				_log.Log(LOG_ERROR,"Invalid Switch command : %s ", Cmd.c_str() );
+		}
+        else
+ 		    _log.Log(LOG_ERROR,"Invalid Switch command : %s ", Cmd.c_str() );
+	}
+}
+
+//get level from imput cmd : selectof value 
+int translateCmd(std::string &Cmd,TOptionMap & options)
+{
+int Level=-1;
```

Comment:
Please check the indent of your source code (in visual studio ctrl-k, ctrl-d)

Reply:
ok  done